### PR TITLE
resouce_manager:  impl resource controller for tokens client

### PR DESCRIPTION
**Pull request by CabinfeverB**
Signed-off-by: Cabinfever_B <cabinfeveroier@gmail.com>

### What problem does this PR solve?

Issue Number: ref #5851

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/mcs/resource_manager/tenant_client/limiter.go [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,g
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tenantclient
```
Comment (nolouch) [inline pkg/mcs/resource_manager/tenant_client/limiter.go:19]: no need introduce tenant,  we only need to introduce resource group.

File: pkg/mcs/resource_manager/tenant_client/client.go [hunk 2]
```diff
@@ -0,0 +1,863 @@
...
+	setupNotificationThreshold float64
+	setupNotificationTimer     *time.Timer
+
+	lastDeadline time.Time
+	lastRate     float64
+
+	limiter *Limiter
+}
+
+func newGroupCostController(ctx context.Context, group *rmpb.ResourceGroup, mainCfg *Config, lowRUNotifyChan chan struct{}) *groupCostController {
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/tenant_client/client.go:360]: It seems that this  `ctx` is unused.

File: pkg/mcs/resource_manager/tenant_client/client.go [hunk 3]
```diff
@@ -0,0 +1,863 @@
...
+	gc.mu.resourceConsumptions = make([]*rmpb.ResourceItem, resourceLen)
+	for typ := range gc.mu.resourceConsumptions {
+		gc.mu.resourceConsumptions[typ] = &rmpb.ResourceItem{
+			Type: rmpb.ResourceType(typ),
+		}
+	}
+	return gc
+}
+
+func (gc *groupCostController) initRunState(ctx context.Context) {
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/tenant_client/client.go:392]: Unused `ctx`.

File: pkg/mcs/resource_manager/tenant_client/limiter.go [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+	lim.notifyThreshold = 0
+	select {
+	case lim.lowTokensNotifyChan <- struct{}{}:
+	default:
+	}
+}
+
+// maybeNotify checks if it's time to send the notification and if so, performs
+// the notification.
+func (lim *Limiter) maybeNotify(now time.Time) {
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/tenant_client/limiter.go:340]: What's the use of the `now`?

File: pkg/mcs/resource_manager/tenant_client/client.go [hunk 5]
```diff
@@ -0,0 +1,863 @@
...
+		responseChan:    make(chan []*rmpb.TokenBucketResponse, 1),
+	}, nil
+}
+
+func (c *resourceGroupsController) Start(ctx context.Context, instanceFingerprint string) error {
+	if len(instanceFingerprint) == 0 {
+		return errors.New("invalid SQLInstanceID")
+	}
+	c.instanceFingerprint = instanceFingerprint
+	// just for demo
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/tenant_client/client.go:99]: Should we stll keep this?

Comment (nolouch) [reply pkg/mcs/resource_manager/tenant_client/client.go:99]: I think should remove it. will add default group later, current should create new group and then alter user to this group.

File: pkg/mcs/resource_manager/tenant_client/client.go [hunk 6]
```diff
@@ -0,0 +1,863 @@
...
+
+type TenantSideKVInterceptor interface {
+	OnRequestWait(ctx context.Context, resourceGroupName string, info RequestInfo) error
+	OnResponse(ctx context.Context, resourceGroupName string, req RequestInfo, resp ResponseInfo) error
+}
+
+type ResourceGroupProvider interface {
+	ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error)
+	GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error)
+	AddResourceGroup(ctx context.Context, resourceGroupName string, settings *rmpb.GroupSettings) (string, error)
```
Comment (nolouch) [inline pkg/mcs/resource_manager/tenant_client/client.go:36]: need consist with the client interface.

File: pkg/mcs/resource_manager/tenant_client/client.go [hunk 7]
```diff
@@ -0,0 +1,863 @@
...
+		if !ok {
+			log.Warn("A non-existent resource group was found when handle token response.", zap.String("name", name))
+		}
+		gc := v.(*groupCostController)
+		gc.handleTokenBucketResponse(ctx, res)
+	}
+}
+
+func (c *resourceGroupsController) collectTokenBucketRequests(ctx context.Context, source string, low bool) {
+	requests := make([]*rmpb.TokenBucketRequst, 0)
```
Comment (nolouch) [inline pkg/mcs/resource_manager/tenant_client/client.go:196]: the `TokenBucketRequst` is typo, kvproto already fix it.

File: pkg/mcs/resource_manager/tenant_client/client.go [hunk 8]
```diff
@@ -0,0 +1,863 @@
...
+	}
+	gc := tmp.(*groupCostController)
+	gc.OnResponse(ctx, req, resp)
+	return nil
+}
+
+type groupCostController struct {
+	resourceGroupName string
+	mainCfg           *Config
+	groupSettings     *rmpb.GroupSettings
```
Comment (nolouch) [inline pkg/mcs/resource_manager/tenant_client/client.go:304]: No GroupSettings.

File: pkg/mcs/resource_manager/client/limiter.go [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+	lim.notifyThreshold = 0
+	select {
+	case lim.lowTokensNotifyChan <- struct{}{}:
+	default:
+	}
+}
+
+// maybeNotify checks if it's time to send the notification and if so, performs
+// the notification.
+func (lim *Limiter) maybeNotify(now time.Time) {
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/limiter.go:291]: no used?

File: pkg/mcs/resource_manager/client/client.go [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+	ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error)
+	GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error)
+	AddResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
+	ModifyResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
+	DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error)
+	AcquireTokenBuckets(ctx context.Context, request *rmpb.TokenBucketsRequest) ([]*rmpb.TokenBucketResponse, error)
+}
+
+func NewResourceGroupController(
+	clientUniqueId uint64,
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/client.go:45]:
```suggestion
	clientUniqueID uint64,
```

File: pkg/mcs/resource_manager/client/client.go [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+			if c.run.requestNeedsRetry || c.shouldReportConsumption() {
+				c.run.requestNeedsRetry = false
+				c.collectTokenBucketRequests(ctx, "report", false /* select all */)
+			}
+		case <-c.lowTokenNotifyChan:
+			c.updateRunState(ctx)
+			c.updateAvgRequestResourcePerSec(ctx)
+			if !c.run.requestInProgress {
+				c.collectTokenBucketRequests(ctx, "low_ru", false /* only select low tokens resource group */)
+				//c.collectTokenBucketRequests(ctx, "low_ru", true /* only select low tokens resource group */)
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/client.go:290]: remove this line?

File: pkg/mcs/resource_manager/client/client.go [hunk 12]
```diff
@@ -0,0 +1,766 @@
...
+	setupNotificationThreshold float64
+	setupNotificationTimer     *time.Timer
+
+	lastDeadline time.Time
+	lastRate     float64
+
+	limiter *Limiter
+}
+
+func newGroupCostController(ctx context.Context, group *rmpb.ResourceGroup, mainCfg *Config, lowRUNotifyChan chan struct{}) *groupCostController {
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/client.go:389]: need use the context?

File: pkg/mcs/resource_manager/client/client.go [hunk 13]
```diff
@@ -0,0 +1,774 @@
...
+		counter.setupNotificationTimer = nil
+		counter.setupNotificationCh = nil
+	}
+	notifyThreshold := granted * notifyFraction
+	if notifyThreshold < bufferRUs {
+		notifyThreshold = bufferRUs
+	}
+
+	var cfg tokenBucketReconfigureArgs
+	if trickleTimeMs == 0 {
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/client.go:618]: maybe we need some comments, here is main difference.

**Review by nolouch — COMMENTED** [review]
rest lgtm

File: pkg/mcs/resource_manager/client/client.go [hunk 14]
```diff
@@ -0,0 +1,774 @@
...
+		provider:           provider,
+		config:             config,
+		lowTokenNotifyChan: make(chan struct{}, 1),
+		tokenResponseChan:  make(chan []*rmpb.TokenBucketResponse, 1),
+		calculators:        []ResourceCalculator{newKVCalculator(config), newSQLCalculator(config)},
+	}, nil
+}
+
+// Start starts resourceGroupController service
+func (c *resourceGroupsController) Start(ctx context.Context) error {
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/client.go:111]: How about add a `Stop` function, I may use it when resource group features disable.

Comment (BornChanger) [reply pkg/mcs/resource_manager/client/client.go:111]: I like it

File: pkg/mcs/resource_manager/client/client.go [hunk 15]
```diff
@@ -0,0 +1,774 @@
...
+			if err = WaitReservations(ctx, now, res); err == nil {
+				break retryLoop
+			}
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if err != nil {
+		return err
+	}
+	gc.mu.Lock()
```
[the remark this answers is not in the document]
Comment (CabinfeverB) [reply pkg/mcs/resource_manager/client/client.go:761]: We can compare it in another PR.

File: pkg/mcs/resource_manager/client/client.go [hunk 16]
```diff
@@ -0,0 +1,790 @@
...
+func (c *resourceGroupsController) sendTokenBucketRequests(ctx context.Context, requests []*rmpb.TokenBucketRequest, source string) {
+	now := time.Now()
+	c.run.lastRequestTime = now
+	c.run.requestInProgress = true
+	req := &rmpb.TokenBucketsRequest{
+		Requests:              requests,
+		TargetRequestPeriodMs: uint64(c.config.targetPeriod / time.Millisecond),
+	}
+	go func() {
+		log.Info("[resource group controllor] send token bucket request", zap.Time("now", now), zap.Any("req", req.Requests), zap.String("source", source))
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/client.go:252]: Debug level.

File: pkg/mcs/resource_manager/client/client.go [hunk 17]
```diff
@@ -0,0 +1,790 @@
...
+	GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error)
+	AddResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
+	ModifyResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
+	DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error)
+	AcquireTokenBuckets(ctx context.Context, request *rmpb.TokenBucketsRequest) ([]*rmpb.TokenBucketResponse, error)
+}
+
+var _ ResourceGroupKVInterceptor = (*resourceGroupsController)(nil)
+
+type resourceGroupsController struct {
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/client.go:56]:
```suggestion
type ResourceGroupsController struct {
```

File: pkg/mcs/resource_manager/client/client.go [hunk 18]
```diff
@@ -0,0 +1,790 @@
...
+	c.groupsController.Store(group.GetName(), gc)
+	return gc, nil
+}
+
+func (c *ResourceGroupsController) updateAllResourceGroups(ctx context.Context) error {
+	groups, err := c.provider.ListResourceGroups(ctx)
+	if err != nil {
+		return err
+	}
+	lastedGroups := make(map[string]struct{})
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/client.go:150]:
```suggestion
	latestGroups := make(map[string]struct{})
```

File: pkg/mcs/resource_manager/client/client.go [hunk 19]
```diff
@@ -0,0 +1,790 @@
...
+		return true
+	})
+}
+
+func (c *ResourceGroupsController) handleTokenBucketResponse(resp []*rmpb.TokenBucketResponse) {
+	for _, res := range resp {
+		name := res.GetResourceGroupName()
+		v, ok := c.groupsController.Load(name)
+		if !ok {
+			log.Warn("A non-existent resource group was found when handle token response.", zap.String("name", name))
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/client.go:221]: Shouldn't we continue here?

**Review by nolouch — APPROVED** [review]
lgtm

File: pkg/mcs/resource_manager/client/client.go (hunk 18 above)
Comment (nolouch) [reply pkg/mcs/resource_manager/client/client.go:153]: reaming @CabinfeverB

**Review by BornChanger — APPROVED** [review]

File: pkg/mcs/resource_manager/client/client.go [hunk 20]
```diff
@@ -0,0 +1,796 @@
...
+			c.handleTokenBucketTrickEvent(ctx)
+		}
+	}
+}
+
+// OnRequestWait is used to check whether resource group has enough tokens. It maybe needs wait some time.
+func (c *ResourceGroupsController) OnRequestWait(
+	ctx context.Context, resourceGroupName string, info RequestInfo,
+) (err error) {
+	if _, ok := defaultWhiteList[resourceGroupName]; ok {
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/client.go:322]: This check could be done by https://github.com/tikv/client-go/pull/648/files#diff-34812587e42268a5d84933be286601fc58fdff17ab336b09ba75455ffe5ed447R96. I prefer keeping it there.

File: pkg/mcs/resource_manager/client/client.go [hunk 21]
```diff
@@ -0,0 +1,796 @@
...
+			return errors.Errorf("[resource group] resourceGroupName %s is not existed.", resourceGroupName)
+		}
+	}
+	err = gc.onRequestWait(ctx, info)
+	return err
+}
+
+// OnResponse is used to consume tokens atfer receiving response
+func (c *ResourceGroupsController) OnResponse(_ context.Context, resourceGroupName string, req RequestInfo, resp ResponseInfo) error {
+	if _, ok := defaultWhiteList[resourceGroupName]; ok {
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/client.go:340]: Ditto.

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

**Review by JmPotato — APPROVED** [review]

**Pull request merged**